lua/mini/ai.lua

```diff
@@ -1562,24 +1562,31 @@ H.get_matched_ranges_builtin = function(captures)
   -- Get parser (LanguageTree) at cursor (important for injected languages)
   local pos = vim.api.nvim_win_get_cursor(0)
   local lang_tree = parser:language_for_range({ pos[1] - 1, pos[2], pos[1] - 1, pos[2] })
-  local lang = lang_tree:lang()
-
-  -- Get query file depending on the local language
-  local query = vim.treesitter.query.get(lang, 'textobjects')
-  if query == nil then H.error_treesitter('query', lang) end
-
-  -- Compute ranges of matched captures
-  local capture_is_requested = vim.tbl_map(function(c) return vim.tbl_contains(captures, '@' .. c) end, query.captures)
 
+  local current = lang_tree
   local res = {}
-  for _, tree in ipairs(lang_tree:trees()) do
-    -- TODO: Remove `opts.all`after compatibility with Neovim=0.10 is dropped
-    for _, match, metadata in query:iter_matches(tree:root(), buf_id, nil, nil, { all = true }) do
-      for capture_id, nodes in pairs(match) do
-        local mt = metadata[capture_id]
-        if capture_is_requested[capture_id] then table.insert(res, H.get_nodes_range_builtin(nodes, buf_id, mt)) end
+  while vim.tbl_isempty(res) and current ~= nil do
+    local lang = current:lang()
+    -- Get query file depending on the local language
+    local query = vim.treesitter.query.get(lang, 'textobjects')
+    if query ~= nil then
+      -- Compute ranges of matched captures
+      local capture_is_requested = vim.tbl_map(
+        function(c) return vim.tbl_contains(captures, '@' .. c) end,
+        query.captures
+      )
+
+      for _, tree in ipairs(current:trees()) do
+        -- TODO: Remove `opts.all`after compatibility with Neovim=0.10 is dropped
+        for _, match, metadata in query:iter_matches(tree:root(), buf_id, nil, nil, { all = true }) do
+          for capture_id, nodes in pairs(match) do
+            local mt = metadata[capture_id]
+            if capture_is_requested[capture_id] then table.insert(res, H.get_nodes_range_builtin(nodes, buf_id, mt)) end
+          end
+        end
```

Comment:
This is too much nesting. Could you, please, extract into a separate helper. Something like `H.append_ranges(ranges, buf_id, query, captures, lang_tree)` that updates `ranges` (i.e. `res` here) in place.